external/kinesis-asl/src/main/scala/org/apache/spark/streaming/kinesis/KinesisDStream.scala

```diff
@@ -0,0 +1,559 @@
+        awsKeypair = Option(awsAccessKeyId, awsSecretKey)))
+      this
+    }
+
+    /**
+     * Create a new instance of [[KinesisInputDStream]] with configured parameters.
+     *
+     * @return Instance of [[KinesisInputDStream]] constructed with configured parameters
+     */
+    def build(): KinesisInputDStream[T] = new KinesisInputDStream(
+      ssc,
+      streamName,
+      endpointUrl,
+      regionName,
+      initialPositionInStream,
+      checkpointAppName,
+      checkpointInterval,
+      storageLevel,
+      messageHandler,
+      kinesisCredsProvider,
+      dynamoDBCredsProvider,
+      cloudWatchCredsProvider)
+
+    private def buildSTSCredentials(
+        stsRoleArn: String,
+        stsSessionName: String,
+        stsExternalId: Option[String] = None,
+        awsKeypair: Option[(String, String)] = None): STSCredentialsProvider = {
+      val baseCredentials = STSCredentialsProvider(
+        stsRoleArn = stsRoleArn,
+        stsSessionName = stsSessionName,
+        stsExternalId = stsExternalId)
+      awsKeypair match {
+        case Some((accessKey, secretKey)) =>
+          baseCredentials.copy(
+            longLivedCredsProvider = BasicCredentialsProvider(accessKey, secretKey))
+        case None =>
+          baseCredentials
+      }
+    }
+  }
+
+  /**
+   * Creates a [[KinesisDStream.Builder]] for constructing [[KinesisInputDStream]] instances
+   * of type [[T]] using the provided message handler.
+   *
+   * @since 2.2.0
+   *
+   * @param ssc StreamingContext to construct DStream with
+   * @param streamName Name of Kinesis stream to read from
+   * @param messageHandler Handler used to extract type [[T]] from a Kinesis [[Record]]
+   * @return [[KinesisDStream.Builder]] instance
+   */
+  def builder[T: ClassTag](
+      ssc: StreamingContext,
+      streamName: String,
+      messageHandler: Record => T): Builder[T] =
+    new Builder(ssc, streamName, messageHandler)
+
+  /**
+   * Creates a [[KinesisDStream.Builder]] for constructing [[KinesisInputDStream]] instances
+   * of type [[Array[Byte]].
+   *
+   * @since 2.2.0
+   *
+   * @param ssc StreamingContext to construct DStream with
+   * @param streamName Name of Kinesis stream to read from
+   * @return [[KinesisDStream.Builder]] instance
+   */
+  def builder(ssc: StreamingContext, streamName: String): Builder[Array[Byte]]
+    = builder(ssc, streamName, KinesisUtils.defaultMessageHandler(_))
+
+  /**
+   * Creates a [[KinesisDStream.Builder]] for constructing [[KinesisInputDStream]] instances
+   * of type [[T]] using the provided message handler.
+   *
+   * @since 2.2.0
+   *
+   * @param jssc JavaStreamingContext to construct DStream with
+   * @param streamName Name of Kinesis stream to read from
+   * @param messageHandler Handler used to extract type [[T]] from a Kinesis [[Record]]
+   * @return [[KinesisDStream.Builder]] instance
+   */
+  def builder[T: ClassTag](
+      jssc: JavaStreamingContext,
```

Comment:
Have the builder take `setStreamingContext`, one which takes `StreamingContext` and the other `JavaStreamingContext`